src/Npgsql/Internal/TypeHandlers/ArrayHandler.cs

```diff
@@ -201,72 +196,75 @@ protected async ValueTask<List<TRequestedElement>> ReadList<TRequestedElement>(N
         internal static class ElementTypeInfo<TElement>
         {
             public static readonly bool IsNonNullable =
-                typeof(TElement).IsValueType && Nullable.GetUnderlyingType(typeof(TElement)) is null;
+                typeof(TElement).IsValueType && default(TElement) is not null;
 
             public static readonly Type NullableElementType = IsNonNullable
                 ? typeof(Nullable<>).MakeGenericType(typeof(TElement))
                 : typeof(TElement);
         }
 
-        internal static class ArrayTypeInfo<TArrayOrList>
+        internal abstract class ArrayTypeInfo<TArray>
         {
             // ReSharper disable StaticMemberInGenericType
-            public static readonly bool IsArray;
-            public static readonly bool IsList;
-            public static readonly Type? ElementType;
-
-            public static readonly Func<ArrayHandler, NpgsqlReadBuffer, bool, ValueTask<Array>> ReadArrayFunc = default!;
-            public static readonly Func<ArrayHandler, NpgsqlReadBuffer, bool, ValueTask<TArrayOrList>> ReadListFunc = default!;
+            public static readonly Type? ElementType = typeof(TArray).IsArray ? typeof(TArray).GetElementType() : null;
             // ReSharper restore StaticMemberInGenericType
 
-            public static bool IsArrayOrList => IsArray || IsList;
+            public static bool IsArray => ElementType is not null;
 
-            static ArrayTypeInfo()
+            static ArrayTypeInfo<TArray>? _derivedInstance;
+            static ArrayTypeInfo<TArray> DerivedInstance
             {
-                var type = typeof(TArrayOrList);
-                IsArray = type.IsArray;
-                IsList = type.IsGenericType && type.GetGenericTypeDefinition() == typeof(List<>);
-
-                ElementType = IsArray
-                    ? type.GetElementType()
-                    : IsList
-                        ? type.GetGenericArguments()[0]
-                        : null;
+                get
+                {
+                    return _derivedInstance ?? CreateInstance();
+                    static ArrayTypeInfo<TArray> CreateInstance()
+                    {
+                        if (ElementType is null)
+                            return null!;
```

Comment:
Can just Debug.Assert here no? Since callers should always check IsArray before?

If so, can move the assertion outside and replace the rest with a single `return _derivedInstance ??= Activator.CreateInstance...`

Reply:
Yup if we say that's the invariant, I tried to mirror the old pattern here somewhat. Will change.